### ISSUE TITLE
[Feature request] Enabling padding-free training with FlexAttention

### ISSUE THREAD

**Issue by iiLaurens** [pytorch/pytorch#140764]
### 🚀 The feature, motivation and pitch

If you really squint your eyes, batch training/inference just a very long sequence of tokens with special form of attention. That is, you can concatenate variably sized sequences of a batch along one dimension and eliminate the batch dimension by only allowing tokens from the same original sequence to attend to each other. This can be achieved with a block-diagonal mask, as shown in the FlexAttention blog. The benefit of doing this is that the model can completely get rid of padding tokens/masks, a.k.a. **padding-free training (PFT)**.

![fg11](https://github.com/user-attachments/assets/af9f2afa-eaa2-4255-af23-a03a51170a9d)

The one thing that prevents FlexAttention from being used for PFT is the softmax operation, Softmax normalizes to 1 with respect to attention scores of all tokens in the sequence. It would be really nice if the softmax was also made more flexible to normalize with respect to a subset of tokens. So suppose have a vector $c_i$ that represents for each token the sequence id it belongs to and $z_i$ is the calculated attention score. Then a softmax variant that would enable PFT would be:

$$s_i = \frac{e^{z_{i}}}{\sum\limits_{c_j = c_i} e^{z_{j}}} \ \ \ for\ i=1,2,\dots,K$$

Alternatively, if flex attention was able to return a vector of softmax values (e.g. multiplying with V), then one could implement a manual renormalization to achieve the same effect. But afaik the current implimentation of of FlexAttention has multiplication with V fused into it.

Do you think supporting this in FlexAttention would be possible and add some value? From my observations, eliminating padding tokens from batch training reduces wasted computations and also allows training/inference with variable-sized sequences.

cc @ezyang @chauhang @penguinwu @zou3519 @bdhirsh @yf225 @Chillee @drisspg @yanboliang @BoyuanFeng

**Comment by iiLaurens** [comment]
I might have had a brain fart indeed. I was still thinking of a one dimensional softmax (because in my mind, batch dimension disappeared) , but there will be a 2d softmax. I think you're right. Will close this!
